gpkit/nomial_data.py

```diff
@@ -0,0 +1,128 @@
+import numpy as np
+
+from collections import defaultdict
+
+from functools import reduce as functools_reduce
+from operator import add
+
+from .small_classes import HashVector, Quantity
+
+from .small_scripts import mag
+
+
+class NomialData(object):
+
+    def __init__(self, exps=None, cs=None, signomials=None, simplify=True):
+        if signomials:
+            exps = functools_reduce(add, (s.exps for s in signomials))
+            cs = np.hstack((mag(s.cs) for s in signomials))
+        if exps is None or cs is None:
+            raise ValueError("creation of a NomialData requires exps and cs.")
+        if simplify:
+            exps, cs = sort_and_simplify(exps, cs)
+        self.exps, self.cs = exps, cs
+        self.varlocs, self.varstrs = locate_vars(self.exps)
+        self.values = {vk: vk.descr["value"] for vk in self.varlocs
+                       if "value" in vk.descr}
+        self._hash = hash(tuple(zip(self.exps, tuple(self.cs))))
+
+    def __hash__(self):
+        return self._hash
+
+    def __repr__(self):
+        return "gpkit.%s(%s)" % (self.__class__.__name__, str(self._hash))
+
+
+def sort_and_simplify(exps, cs, return_map=False):
+    """Reduces the number of monomials, and casts them to a sorted form.
+
+    Arguments
+    ---------
+
+    exps : list of Hashvectors
+        The exponents of each monomial
+    cs : array of floats or Quantities
+        The coefficients of each monomial
+    return_map : bool (optional)
+        Whether to return the map of which monomials combined to form a
+        simpler monomial, and their fractions of that monomial's final c.
```

Comment:
left-field idea -- to deal with the fact that `matches` computation is shared between exps/cs and maps

 - make `matches` a property of `NomialData` (with possibly an updated name). Implement by initializing matches attribute to `None` in `NomialData.__init__`, and recomputing matches only if thy are `None`.
 - Use matches to implement the computing of exps and cs from sort_and_simplify
 - separate mmaps function that also uses `matches`